src/esrs/postgres/event_store.rs

```diff
@@ -0,0 +1,296 @@
+                .into_iter()
+                .map(|event| Ok(event.try_into()?))
+                .collect::<Result<Vec<StoreEvent<A::Event>>, Self::Error>>()?,
+        )
+    }
+
+    #[tracing::instrument(skip_all, fields(aggregate_id = % aggregate_state.id()), err)]
+    async fn persist(
+        &self,
+        aggregate_state: &mut AggregateState<A::State>,
+        events: Vec<A::Event>,
+    ) -> Result<Vec<StoreEvent<A::Event>>, Self::Error> {
+        let mut transaction: Transaction<Postgres> = self.inner.pool.begin().await?;
+        let occurred_on: DateTime<Utc> = Utc::now();
+        let mut store_events: Vec<StoreEvent<A::Event>> = vec![];
+
+        let aggregate_id = *aggregate_state.id();
+
+        for event in events.into_iter() {
+            let store_event: StoreEvent<<A as Aggregate>::Event> = self
+                .save_event(
+                    aggregate_id,
+                    event,
+                    occurred_on,
+                    aggregate_state.next_sequence_number(),
+                    &mut *transaction,
+                )
+                .await?;
+
+            store_events.push(store_event);
+        }
+
+        // Acquiring the list of transactional event handlers early, as it is an expensive operation.
+        let transactional_event_handlers = &self.inner.transactional_event_handlers;
+        for store_event in &store_events {
+            for transactional_event_handler in transactional_event_handlers.iter() {
+                let span = tracing::trace_span!(
+                    "esrs.transactional_event_handler",
+                    event_id = %store_event.id,
+                    aggregate_id = %store_event.aggregate_id,
+                    transactional_event_handler = transactional_event_handler.name()
+                );
+                let _e = span.enter();
+
+                if let Err(error) = transactional_event_handler.handle(store_event, &mut transaction).await {
+                    tracing::error!({
+                        event_id = %store_event.id,
+                        aggregate_id = %store_event.aggregate_id,
+                        transactional_event_handler = transactional_event_handler.name(),
+                        error = ?error,
+                    }, "transactional event handler failed to handle event");
+
+                    return Err(error);
+                }
+            }
+        }
+
+        transaction.commit().await?;
+
+        // We need to drop the lock on the aggregate state here as:
+        // 1. the events have already been persisted, hence the DB has the latest aggregate;
+        // 2. the event handlers below might need to access this aggregate atomically (causing a deadlock!).
+        drop(aggregate_state.take_lock());
+
+        // Acquiring the list of event handlers early, as it is an expensive operation.
+        let event_handlers = self.inner.event_handlers.read().await;
+        for store_event in &store_events {
```

Comment:
But i think it's the same. It's an expensive operation since it's getting a lock from the RwLock.. No?

Reply:
Yeah but it's not "acquiring the list EARLY" (stress on **early**). That was the case before because we were ArcSwapping in a loop, which was bad.
Now you are forced to acquire the lock beforehand, there's no way around it (I guess you can obtain the lock inside the loop every time, but that's visibly stupid, unlike the old ArcSwap that had no clear semantics).
Yeah whatever this is borderline fine I guess.